Add `lint-v2` rule and implement it for Python with black.

pierrechevalier83 opened the pull request:
### Problem

In the near future, we will want to add a pre-commit hook that helps us remember to automatically format our code with black.
Currently, the version of black that's integrated in pants will always modify the user's code, which may not be desirable in a pre_commit hook situation.

### Solution

Add a `lint-v2` rule which uses black with the `--check` flag to highlight unformatted files without actually modifying them.

### Result

See integration tests for examples: User can call `./pants lint-v2 ...`. The command will fail if the files would be formatted, but won't rewrite the files.

illicitonion commented on src/python/pants/backend/python/subsystems/black.py:
```diff
@@ -16,3 +16,5 @@ def register_options(cls, register):
     super().register_options(register)
     register('--config', advanced=True, type=file_option, fingerprint=True,
               help="Path to formatting tool's config file")
+    register('--check', advanced=True, type=bool, fingerprint=True,
```
Long-term, there's an interesting question of direction here.

Historically, we had the `lint` goal, which is basically `fmt-without-formatting` - in the future, would we like:
1. A `lint` goal alongside the `fmt` goal?
2. A `fmt`-wide flag called something like `--check`
3. A per-formatter flag called `--check`

I suspect long-term we want 1 or 2, and we should work that out before we name the v2 console_rule `fmt`, but while this is experimental until then, this works :)

stuhood commented:
> The code here looks good to me, but I'm still unsure of the long-term direction; I think that ignoring history, I probably would design these goals as `fmt` an `fmt --check` rather than `fmt` and `lint` - maybe this isn't where we want to make that change, but I think it's worth a discussion.

The issue with `./pants fmt --check` is that you probably also want to check other lints that are not formatting related. So then you need to run `./pants fmt --check lint`. Maybe that's fine, and the distinction we draw is that only autoformatters go in `fmt`, and that the style checking that they represent is sufficiently distinct from other lints to be worth having split out?

My feeling though is that all realistic cases will want to run "all of the lints" (assuming that they are fast enough).

benjyw commented:
Thanks - code looks good. After some thought, I come down on the side of putting this in a `lint` goal.  For the reason Stu stated - we may have other linters that are not tied to formatters. And it makes more sense to run `./pants lint` than `./pants lint fmt --check` in order to get all the lint checks.

Another way to think of it is that "formatter in check mode" is an implementation detail of a lint rule.

Eric-Arellano commented:
+1 to this being a lint goal. 

(Haven’t checked the code at all.)

pierrechevalier83 commented on src/python/pants/rules/core/lint.py:
```diff
@@ -0,0 +1,67 @@
+    if result.stderr:
+      console.print_stderr(result.stderr)
+    if result.exit_code != 0:
+      yield Lint(result.exit_code)
```
I added a couple of tests: lint multiple files and lint multiple targets. With the latter, I could reproduce the issue you're describing (we don't report all failing targets). I fixed it as you described.

pierrechevalier83 commented on src/python/pants/rules/core/lint.py:
```diff
@@ -0,0 +1,67 @@
+  results = yield [
+          Get(LintResult, FmtTarget, target.adaptor)
+          for target in targets
+          # @union assumes that all targets passed implement the union, so we manually
```
Will address in a separate PR to hopefully get this one merged today (I'm also dragging https://github.com/pantsbuild/pants/pull/8454 which is based on top of this, so would be keen on iterating fast if possible)

pierrechevalier83 replied:
Done here: https://github.com/pantsbuild/pants/pull/8490

pierrechevalier83 commented:
Just a comment to be explicit: the reasoning for `lint` vs `fmt --check` makes sense to me. I prepped the `lint` version of the PR for merging (rebased + addressed review comments)

benjyw commented on src/python/pants/backend/python/rules/python_fmt.py:
```diff
@@ -22,6 +26,7 @@
 from pants.engine.rules import UnionRule, optionable_rule, rule
 from pants.engine.selectors import Get
 from pants.rules.core.fmt import FmtResult, FmtTarget
```
I think FmtTarget should be renamed TargetWithSources or something like that. Now that we're using it outside the Fmt context.

pierrechevalier83 replied:
Addressed in new commit in this PR.

pierrechevalier83 commented:
The fight against flakey CI has been too intense. I'm reducing from 3 PRs to one to increase my likelihood of ever getting my code merged.
Closing in favour of https://github.com/pantsbuild/pants/pull/8553 which contains https://github.com/pantsbuild/pants/pull/8452, https://github.com/pantsbuild/pants/pull/8454 and https://github.com/pantsbuild/pants/pull/8490
